File: Ray Tracing next Week/util/week1/interval.py
# util/interval.py

class Interval:

    def __init__(self, min_val=float("inf"), max_val=float("-inf")):
        """
        Default = empty interval
        """
        self.min = min_val
        self.max = max_val

    def size(self):
        """
        Length of interval
        Formula:
        size = max - min
        """
        return self.max - self.min

    def contains(self, x):
        """
        Inclusive:
        min <= x <= max
        """
        return self.min <= x <= self.max

    def surrounds(self, x):
        """
        Strict:
        min < x < max
        """
        return self.min < x < self.max

    def clamp(self, x):
        """
        Clamp x into [min,max]
        """

        if x < self.min:
            return self.min

        if x > self.max:
            return self.max

        return x


# Static intervals
Interval.empty = Interval(float("inf"), float("-inf"))
Interval.universe = Interval(float("-inf"), float("inf"))

File: Ray Tracing in One Weekend - Javascript/Ch10/interval.js
class Interval{

    constructor(min,max){
        this.max = max
        this.min = min
    }

    size(){
        return this.max-this.min
    }

    contains(x){
         return (this.min <= x) && (x <= this.max)
    }

    surrounds(x){
        return (this.min < x) && (x < this.max)
    }

    clamp(x){
        if(x < this.min) return this.min 
        if(x > this.max) return this.max
        return x
    }


    
}

var empty = new Interval(Number.MAX_VALUE,Number.MIN_VALUE)
var universe = new Interval(Number.MIN_VALUE,Number.MAX_VALUE)

window.Empty = empty
window.Universe = universe

window.Interval = Interval